ement.el

```diff
@@ -222,6 +216,51 @@ call `pop-to-buffer'."
 
 ;;;; Functions
 
+(defun ement--read-username-and-server (prompt &optional initial-input history
+                                               default-value
+                                               inherit-input-method)
+  "Read Matrix user ID, submitted in the form \"@username:server\".
+
+Return (user-input . plist) where plist is `auth-source-search'-compatible plist."
+  (cl-loop
+   as user-id = (read-string prompt
+                             initial-input history default-value
+                             inherit-input-method)
+   if (not (string-match
+            (rx bos "@" (group (1+ (not (any ":")))) ; Username
+                ":" (group (optional (1+ (not (any blank)))))) ; Server name
+            user-id))
+   do (minibuffer-message "Invalid user ID format: use @USERNAME:SERVER")
+   else return (cons user-id
+                     (list :user (match-string 1 user-id)
+                           :host (match-string 2 user-id)))))
+
+(defun ement--read-user-credentials (&optional initial-input history
+                                               default-value
+                                               inherit-input-method)
+  "Read Matrix user credentials.  Return list (user-id password).
+
+Optional arguments refer to reading user ID."
+  (cl-destructuring-bind (user-id . plist) (ement--read-username-and-server
+                                            "User ID: "
+                                            initial-input history
+                                            default-value
+                                            inherit-input-method)
+    (list user-id
+          (save-match-data
+            ;; save-match-data is necessary at least when
+            ;; auth-source does not use cache
+            (or (when-let ((auth-source-secret
+                            (when (fboundp 'auth-source-search)
```

Comment:
I don't use auth-source, but is this `fboundp` check really necessary?  Having to use this might indicate that we should be doing something else.

Reply:
I'll drop it.

I tend to write code that would be easy to make optional.  Passwords is a matter delicate enough for this.  It has become a habit but it better be left in my libraries.

Reply:
If we're `require`ing `auth-source`, then the function should be available, right?  So it shouldn't be necessary to test this.

Supporting things optionally in Emacs is usually a bit of a hassle.  I'd prefer to keep the code as simple as possible.